### ISSUE TITLE
We should have a podcast2zim scraper

### ISSUE THREAD

**Issue by Popolechien** [openzim/overview#42]
As discussed in today's meeting. Potentially a GSoC project.

**Comment by aaryansinhaa** [comment]
really like this idea,

From basic googling, i guess the primary distribution for podcasts is RSS feeds, where the XML contains metadata for episodes along with enclosure links pointing to the audio files. If that's the case, should the first target of `podcast2zim` be RSS feeds-> parsing the XML-> downloading the audio files -> generating a static website structure for the podcast inside the ZIM?

Or is the goal to scrape podcast platforms directly (for example Spotify or similar services)? I assume that would be significantly more complex compared to RSS scraping.

WDYT?
@kelson42 @benoit74  @Popolechien @audiodude @rgaudin

**Comment by rgaudin** [comment]
> From basic googling, i guess the primary distribution for podcasts is RSS feeds, where the XML contains metadata for episodes along with enclosure links pointing to the audio files. If that's the case, should the first target of `podcast2zim` be RSS feeds-> parsing the XML-> downloading the audio files -> generating a static website structure for the podcast inside the ZIM?

exactly.

> Or is the goal to scrape podcast platforms directly (for example Spotify or similar services)? I assume that would be significantly more complex compared to RSS scraping.

I believe those platforms are aggregators that consume the actual RSS feeds.
That could be a future enhancement. I quickly checked with the Web version of Apple Podcast (not signed in) and I could retrieve the `feedUrl` from the requests. Being able to paste an Apple Podcast URL to get the ZIM would indeed be a great UX.

**Comment by aaryansinhaa** [comment]
cool idea, would really like to be involved in it.
I checked spotify, it's slightly different from apple, where they internally store the files, and no rss is available, this is probably the case in which creators upload their podcast via spotify for creators, and maybe because most of their audio urls are protected. apple podcast does provide a feedUrl.

regardless, I think that the first iteration should be RSS focused and then we can move to supporting platforms in future versions.

I'll research more and will write a solid proposal. thanks

**Comment by kelson42** [comment]
Atom has probably to be supported too.

**Comment by kelson42** [comment]
In term of software stack/architecture, could we not have (reuse) a pure JS podcast reader? That would simplify a lot the work to do!

**Comment by aaryansinhaa** [comment]
Wanted a clarification, by podcast reader you mean a rss/atom parser, right?
I saw a few parser, I believe the most popular is [feedparser](https://github.com/kurtmckee/feedparser), which is a pure python implementation, I also found a similar [node-feedparser](https://github.com/danmactough/node-feedparser) implementation, which is a JS implementation.

I believe that python's feedparser is our best way forward, as youtube2zim is also following the same pattern, it uses yt-dlp to get metadata and videos, and use existing utils in zimscraperlib. and we can maybe use premade frontend [Podlove Player](https://github.com/podlove/podlove-ui/blob/development/apps/player/README.md).(yt2zim uses a custom vue frontend)


the flow I am thinking is:
rss/atom url -> feedparser -> json based fixtures(chapter.json, episode.json, and some other metadata related) -> frontend from podlove, all this bundled to a zim file.


is this the correct direction? or are you thinking of using a pure js reader that directly consumes rss feed?

**Comment by aaryansinhaa** [comment]
**TLDR:** podcast link -> feedUrl -> feed.py(scraping rss xml) -> download the media files -> generate a static webpage using the json and downloaded audio + transcript ->zim.py(package into a single bundle) -> a podcast in zim.

---

some main flags that I am thinking: 

 **Core**
1. `--name`: name of the zim file
2. ` --feedUrl`: downloads and zims a podcast using rss/atom link.
3. `--concurrency: ` number of files to download concurrently.
4. `--retries: ` number of retries for a failed download.

**Feature**

1. . `--podcastUrl`: downloads a podcast from apple podcasts url and zims it
2. `--episode-limit n`: zims the latest `n` episodes.
3. `--dateAfter: ` zims the episodes release after a particular date(DD-MM-YYYY).

---

Some points:
1. **schemas to store:** podcast, episode and episode preview, transcript.
2. **UI:** like the rough concept shown below:  

<img width="1017" height="570" alt="Image" src="https://github.com/user-attachments/assets/461b5345-7ba1-4d42-bc0b-49b1d0585058" />

> This will be the index page of our zim, the episode will have a separate page, with transcript, player and thumbnail etc.

3. Some of the podcasts  also have a season based release, so I think that it would be a good ux to include that as well, like the ability to zim a particular season of the podcast, idk whether to place in v1 or not, depends on the timeline of this project, and also user feedback if it's necessary. 

4. I also think it is a good idea to support apple music in the v1, like simply pasting the podcast link, and have our scrapper scrape the feedUrl, I don't think it is a very complex problem to solve, since we are able to get feedUrl from requests, and it will also set us up to support more platforms in the future.


@kelson42 @benoit74 @Popolechien @rgaudin I am really awaiting feedback, so please share feedback on my approach.
also please share who will be the mentor in this project, so that I can ping the right person, thanks

**Comment by rgaudin** [comment]
@aaryansinhaa looks good. I also think Apple Podcast should be in first version (and/or another service but this one is easy to handle and is quite popular).
I also think that selection will be key to the success of the scraper. Ability to choose seasons as you said, or maybe create mix/collection of multiple shows. You should include some time in your proposal to study the podcast landscape and find out what would be best for this offline scenario: since the target is not connected people following a single show (all apps allow downloading episodes) but some people preparing ZIM files for a different audience. I imagine thematic collections, etc.
This raises the complexity of the UX obviously but in your proposal you can include different steps and an earlier one can be a single-show scraper for instance.
Duration is up to you. GSoC offers three different duration.

**Comment by hari-om65** [comment]
Hi Kiwix team! 

I'm Hari Om Singh, student at B.I.T. Sindri. I am very interested in the `podcast2zim` scraper project for GSoC 2026.

I have developed a working end-to-end prototype that handles the full pipeline from RSS ingestion to ZIM archival. Key technical milestones achieved in this prototype include:

1. **Robust RSS Parsing:** Implemented `feedparser` to dynamically extract episode metadata, titles, and media enclosures (MP3/AAC) from live podcast feeds.
2. **Asset Management:** Developed an automated downloader using `requests` with local directory management to stage media for packaging.
3. **Advanced ZIM Packaging:** To ensure compatibility with the latest `libzim` (3.8.0) Python bindings, I implemented a custom `Item` class and utilized `FileProvider`. This was necessary to resolve the recent API shift where `add_item` no longer accepts positional arguments, requiring a structured object-oriented approach instead.
4. **Metadata Integrity:** Successfully mapped RSS XML attributes to ZIM metadata (Language, Title, Creator) to ensure the final archive is searchable and compliant with OpenZIM standards.

You can find my functional prototype and detailed technical documentation here: 
https://github.com/hari-om65/podcast2zim-prototype

I am eager to contribute to the Kiwix mission of making educational content accessible offline. I look forward to discussing the full-scale implementation and submitting my formal proposal!

**Comment by aaryansinhaa** [comment]
Hey @rgaudin , completely agree that our scraper serves a different audience. I am thinking of three major phases:

1. **Single Show Scraper:** Covering feedUrl -> zim flow, it would lay a good working ground to start adding features.
```
---name, --feed-url --concurrency --retries --podcast-url --episode-limit --date-after
```
Would all be covered in this phase.

2. **Thematic Compilation:** I'm considering adding a feature, `--compile-json` where we can compile from a json file.
The idea is that a curator creates a JSON file describing a collection:
```json
{
  "title": "True Crime — Curated Compilation",
  "description": "A curated offline collection covering investigations and criminal justice.",
  "podcasts": [
    {
      "name": "Serial",
      "description": "Investigative journalism — Season 1 re-examines the Adnan Syed case.",
      "feed-url": "https://feeds.simplecast.com/xl56lcl8",
      "season": 1
    },
    {
      "name": "Casefile True Crime",
      "description": "Meticulously researched cases from around the world.",
      "feed-url": "https://casefilepodcast.com/feed/podcast",
      "episode-limit": 20,
      "date-after": "2023-01-01"
    }
  ]
}
```

The scraper then fetches, filters, and packages all of these into one ZIM file with a multi-show home page.

3. **Curator Tool:** For the final phase, I think we can provide a simple web based curator tool, which automatically generates a curated json from pasted links and filters; we can integrate this to the scraper, so people don't have to handwrite jsons.


I think making these three as the core deliverables for gsoc project would make sense(medium-long), and I'll start making weekly timeline for my proposal, I'd really appreciate any feedback. thanks.

**Comment by AdeelTahir-SE** [comment]
Hi Kiwix Team,

I found out about Kiwix through the list of participating organizations and really like the idea of making content available even when offline. I really want to contribute to `Podcast2Zim`.
From the research I’ve done, I found out that ideas is:

1. to get podcasts through RSS feeds by parsing and downloading audio along with related metadata, and then converting that into ZIM files, which can be accessed offline.
2. to get podcasts from aggregators like Apple Podcasts, parse them, and convert them into zim files
3. to get podcasts through both.

Could you please verify if my understanding and approach are correct? Specifically, should we focus on both approaches or just one of them? The reason is I am currently working on a prototype as well, and do not want to build the wrong product. I will also share details of the Prototype soon.

Also, please check and verify the expected architecture diagram:

<img width="262" height="582" alt="Image" src="https://github.com/user-attachments/assets/3c658de2-244b-4123-93bb-0bb4434a9e9b" />

@rgaudin @kelson42 @Popolechien

**Comment by AdeelTahir-SE** [comment]
I have implemented the first version : https://github.com/AdeelTahir-SE/podcast2zim
_Current features:_
**Fetch Podcast from RSS Feed( The Daily)** => **Parse the feed** => **Save to Static Website and Media Files** => **Convert to ZIM using `zimscraperlib`**
_Upcoming Features:_

- Support for multiple Podcast Sources
- Improved Website Structure
- More control and selective downloading instead of just specifying number of episodes
- metadata improvements like title and description

here is demo (screenshot) of basic podcast zim file created through above prototype created at kiwix (design needs improvement😅 right now):
**Note: 2 episodes were specified while creating zim file**

![Image](https://github.com/user-attachments/assets/d992d46a-08cf-4db7-9b42-680559f872ce)
![Image](https://github.com/user-attachments/assets/a2e75766-445c-4ecc-bd79-20823b00e244)

**Comment by Madhulathayallamanda** [comment]
Hi, I’m interested in working on podcast2zim for GSoC 2026.
I’ve started building a small prototype to understand the pipeline. Right now it parses RSS feeds, extracts episode details, and I’ve also added basic audio downloading for a few episodes.
Here is my repo: https://github.com/Madhulathayallamanda/podcast2zim-prototype
Next, I’m planning to structure the data and move towards generating ZIM files.
Please let me know if this approach looks fine or if I should focus on anything specific.

**Comment by peniel-09** [comment]
Hi, I'm a GSoC 2026 applicant interested in podcast2zim. One thing that hasn't been discussed yet: transcript support. Most podcast feeds don't include transcripts, but they significantly increase the value of a ZIM: especially for search, accessibility, and offline training contexts. Would it be in scope to optionally generate them at scrape time using a local model like Whisper (--transcribe flag)? Curious if the team has already considered this.

**Comment by Dev-angPatil** [comment]
Hi, I have been going through the discussion and had a thought on the pipeline design.
Since the flow is essentially RSS/Atom → parse → download → generate static site → package into ZIM, would it make sense to structure them as independent modules (parser, downloader, renderer, packager)?

This way:
* RSS/Atom support can be the initial implementation
* Adding Apple Podcasts or other platforms later would just require a new “input adapter”
* Features like multi-show compilation or filtering could reuse the same pipeline without changing core logic

Also, instead of generating static pages first, would it be possible to directly map parsed data into ZIM structures and only use a lightweight template layer for rendering? This could reduce duplication between “site generation” and “packaging”.

Also, building on the transcript idea mentioned above, this could fit well as an optional stage in the pipeline  keeping the core workflow lightweight.
Does this direction make sense, or is the current architecture already optimized for this?

**Comment by mvban** [comment]
@aaryansinhaa I agree with your approach. One suggestion I can make is to implement the podcastindex.org API for podcast discovery.  It's a directory of RSS feeds.

**Comment by mvban** [comment]
@aaryansinhaa I'm curious if we can link transcripts, show notes to the audio file?

**Comment by benoit74** [comment]
From my experience podcastindex.org is not that great.

See for instance:
- https://podcastindex.org/podcast/6098687 which is stuck in 2024 while there new episodes from 2025: https://open.spotify.com/show/2taDRmHcJepbPiaE0R6a0m
- episodes do not play at all (so link to MP3 in API is wrong)

**Comment by aaryansinhaa** [comment]
@mvban afaik, some podcasts have `<podcast:transcript>`, but lately some publishers are using slug(`/transcript`) after episode link to link their transcripts directly inside the platform. This would be an interesting problem to solve, as mostly, you will be hittig 404, on `episode_link/transcript`. also @benoit74 is right, we probably don't want to use podcastindex for feed discovery. I think https://rss.com/community/ is a much more reliable platform. I'd love to hear your views. thanks  

Adding transcripts though would be a really great move for our users.
